openvdb/openvdb/points/PointStatistics.h

```diff
@@ -0,0 +1,609 @@
+        ElementT l = b.first.lengthSqr();
+        if (l < mLengths.first) {
+            mLengths.first = l;
+            mMinMax.first = b.first;
+        }
+        l = b.second.lengthSqr();
+        if (l > mLengths.second) {
+            mLengths.second = l;
+            mMinMax.second = b.second;
+        }
+    }
+
+    std::pair<ElementT, ElementT> mLengths;
+    ExtentT mMinMax;
+};
+
+/// @brief  Vector component-wise op to evaluate the min/max of
+///   vector components and return the result as a vector of
+///   equal size and precision
+template <typename VecT>
+struct ComponentExtent
+{
+    using ExtentT = std::pair<VecT, VecT>;
+    ComponentExtent(const VecT& init) : mMinMax(init, init) {}
+    ComponentExtent(const ExtentT& init) : mMinMax(init) {}
+    inline const ExtentT& get() const { return mMinMax; }
+    inline void operator()(const VecT& b)
+    {
+        mMinMax.first = math::minComponent(mMinMax.first, b);
+        mMinMax.second =  math::maxComponent(mMinMax.second, b);
+    }
+    inline void operator()(const ExtentT& b)
+    {
+        mMinMax.first = math::minComponent(mMinMax.first, b.first);
+        mMinMax.second = math::maxComponent(mMinMax.second, b.second);
+    }
+
+    ExtentT mMinMax;
+};
+
+template <typename ValueT,
+    typename CodecT,
+    typename FilterT,
+    typename ExtentOp,
+    typename PointDataTreeT>
+inline bool evalExtents(const PointDataTreeT& points,
+        const std::string& attribute,
+        typename ExtentOp::ExtentT& ext,
+        const FilterT& filter,
+        typename PointDataTreeT::template ValueConverter
+            <typename ExtentOp::ExtentT::first_type>::Type* const minTree = nullptr,
+        typename PointDataTreeT::template ValueConverter
+            <typename ExtentOp::ExtentT::second_type>::Type* const maxTree = nullptr)
+{
+    static_assert(std::is_base_of<TreeBase, PointDataTreeT>::value,
+        "PointDataTreeT in instantiation of evalExtents is not an openvdb Tree type");
+
+    struct ResultType {
+        typename ExtentOp::ExtentT ext;
+        bool data = false;
+    };
+
+    tree::LeafManager<const PointDataTreeT> manager(points);
+    if (manager.leafCount() == 0) return false;
+    const size_t idx = manager.leaf(0).attributeSet().find(attribute);
+    if (idx == AttributeSet::INVALID_POS) return false;
+
+    // track results per leaf for min/max trees
+    std::vector<std::unique_ptr<typename ExtentOp::ExtentT>> values;
+    if (minTree || maxTree) values.resize(manager.leafCount());
+
+    const ResultType result = tbb::parallel_reduce(manager.leafRange(),
+        ResultType(),
+        [idx, &filter, &values]
+            (const auto& range, ResultType in) -> ResultType
+        {
+            for (auto leaf = range.begin(); leaf; ++leaf) {
+                AttributeHandle<ValueT, CodecT> handle(leaf->constAttributeArray(idx));
+                if (handle.size() == 0) continue;
+                if (std::is_same<FilterT, NullFilter>::value) {
```

Comment:
@danrbailey we really need to talk about point filters. I would love to be able to use what I have here i.e. if a `NullFilter` is used, iterate directly on the array. It is so much faster than using the `IndexFilter` API. Additionally, we never actually conform to the `ActiveFilter`/`BinaryFilter` tools. Functions seem to only iteration over ValueOn data, but surely this is wrong? Tools should default to a `FilterT=ActiveFilter` and should  use `leaf->beginIndexAll()`. Otherwise these filter become redundant. In so doing, the `NullFilter` option provides direct access to doing ` for (size_t i = 1; i < size; ++i)` which is much nicer.

Reply:
@danrbailey be good to chat about this as I feel like it's a considerable limitation in the points API. But I don't think it should stop this going on.

Reply:
Done! I swear when i originally implemented this there was some issue with filter.state() which wasn't giving me the desired results but that's so long ago I might be misremembering. Works great for now and will keep an eye on it.